src/c_api.rs

```diff
@@ -0,0 +1,612 @@
+    path: *const c_char,
+    str: *mut *const c_char,
+    len: *mut libc::size_t,
+) -> c_int {
+    if !key.is_null() {
+        let key = unsafe { &mut *key };
+        let path = unsafe { CStr::from_ptr(path).to_str().unwrap() };
+        if let Ok(s) = key.redis_json.to_string(path, Format::JSON) {
+            if let Ok(s) = CString::new(s) {
+                unsafe {
+                    *str = s.as_bytes_with_nul().as_ptr() as *const c_char;
+                    *len = s.as_bytes().len();
+                }
+                key.cstr_val = Some(s);
+                return RedisReturnCode::REDISMODULE_OK as c_int;
+            }
+        }
+    }
+    RedisReturnCode::REDISMODULE_ERR as c_int
+}
+
+#[no_mangle]
+pub extern "C" fn JSONAPI_getRedisModuleString(
+    json: JSONApiPathRef,
+    str: *mut *mut rawmod::RedisModuleString,
+) -> c_int {
+    if !json.is_null() {
+        let json = unsafe { &*json };
+        if let Some(ref p) = json.path {
+            if let Ok(res) = json.json_key.redis_json.to_string(p, Format::JSON) {
+                return json.json_key.create_rmstring(res.as_str(), str);
+            }
+        }
+    }
+    return RedisReturnCode::REDISMODULE_ERR as c_int;
+}
+
+#[no_mangle]
+pub extern "C" fn JSONAPI_getRedisModuleStringFromKey(
+    key: JSONApiKeyRef,
+    path: *const c_char,
+    str: *mut *mut rawmod::RedisModuleString,
+) -> c_int {
+    if !key.is_null() {
+        let key = unsafe { &*key };
+        let path = unsafe { CStr::from_ptr(path).to_str().unwrap() };
+        if let Ok(res) = key.redis_json.to_string(path, Format::JSON) {
+            key.create_rmstring(res.as_str(), str)
+        } else {
+            RedisReturnCode::REDISMODULE_ERR as c_int
+        }
+    } else {
+        RedisReturnCode::REDISMODULE_ERR as c_int
+    }
+}
+
+fn get_int_value(value: &Value) -> Option<c_long> {
+    match value {
+        Value::Number(ref n) => {
+            if let Some(i) = n.as_i64() {
+                return Some(i);
+            }
+        }
+        Value::String(ref s) => {
+            if let Ok(v) = c_long::from_str(s.as_str()) {
+                return Some(v);
+            }
+        }
+        Value::Bool(ref b) => {
+            if *b {
+                return Some(1);
+            } else {
+                return Some(0);
+            }
+        }
+        _ => {}
+    }
+    return None;
+}
+
+fn get_double_value(value: &Value) -> Option<c_double> {
+    match value {
+        Value::Number(n) => {
+            if let Some(f) = n.as_f64() {
+                return Some(f);
+            }
+        }
+        Value::String(ref s) => {
+            if let Ok(v) = c_double::from_str(s.as_str()) {
+                return Some(v);
+            }
+        }
+        Value::Bool(ref b) => {
+            if *b {
+                return Some(1.0);
+            } else {
+                return Some(0.0);
+            }
+        }
+        _ => {}
+    }
+    return None;
```

Comment:
It would be more idiomatic in Rust to:
1. Get rid of all those `return`s and use a single expression instead.
2. There's no need for these explicit `ref`s in the `match` statement, the compiler infers those by itself.
3. When there's an `if` inside a `match` it can often me combined with the `match`.
4. Standard combinators like `ok()` can convert between `Result` and `Option`, no need for another `match` there.

So all of that code can be simplified to this:
```suggestion
    match value {
        Value::Number(n) if n.is_f64() => n.as_f64(),
        Value::String(s) => c_double::from_str(s.as_str()).ok(),
        Value::Bool(b) => Some(if *b { 1.0 } else { 0.0 }),
        _ => None,
    }
```

Reply:
👍🏼